### ISSUE TITLE
bug: z_getsubtreesbyindex can return 2 subtrees when 1 was requested

### ISSUE THREAD

**Issue by teor2345** [ZcashFoundation/zebra#7708]
### What happened?

 I expected to see this happen:

When 1 subtree is requested, 0 or 1 subtrees are returned.

Instead, this happened: 

Sometimes 2 subtrees are returned, and there is a gap between them.

### What were you doing when the issue happened?

Running tests in CI on a cached state:
https://github.com/ZcashFoundation/zebra/actions/runs/6453229873/job/17516699670?pr=7703#step:9:355
https://github.com/ZcashFoundation/zebra/actions/runs/6453229873/job/17516707077?pr=7703#step:9:302

#### Snapshot test

```
Snapshot file: zebrad/tests/snapshots/z_getsubtreesbyindex_mainnet_sapling_0_1.snap
Snapshot: z_getsubtreesbyindex_mainnet_sapling_0_1
Source: zebrad/tests/acceptance.rs:2723
────────────────────────────────────────────────────────────────────────────────
Expression: parsed
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
    5     5 │     "subtrees": [
    6     6 │       {
    7     7 │         "root": "754bb593ea42d231a7ddf367640f09bbf59dc00f2c1d2003cc340e0c016b5b13",
    8     8 │         "end_height": 558822
          9 │+      },
         10 │+      {
         11 │+        "root": "ffdd1a0628296bf274daa5ea13fb57818d678957d49f91ccd69efae0bac24934",
         12 │+        "end_height": 2255074
    9    13 │       }
   10    14 │     ]
   11    15 │   },
```

#### gRPC test

```
The application panicked (crashed).
Message:  The response from the `z_getsubtreesbyindex` RPC contains a wrong number of Sapling subtrees.
Location: zebrad/tests/common/lightwalletd/wallet_grpc_test.rs:429
```

### Zebra Version

zebrad 1.2.0 PR #7703

### Which operating systems does the issue happen on?

- [X] Linux
- [ ] macOS
- [ ] Windows
- [ ] Other OS

### OS details

CI

### Additional information

This seems to happen when there is a new subtree in the non-finalised state, so the bug is probably in the state code that merges finalised and non-finalised subtrees.

**Comment by teor2345** [comment]
@mpguerra this looks like a blocker for releasing the subtree RPCs, let's schedule it soon?

**Comment by upbqdn** [comment]
> This seems to happen when there is a new subtree in the non-finalised state, so the bug is probably in the state code that merges finalised and non-finalised subtrees.

I wonder how you figured this out.

**Comment by teor2345** [comment]
> > This seems to happen when there is a new subtree in the non-finalised state, so the bug is probably in the state code that merges finalised and non-finalised subtrees.
> 
> I wonder how you figured this out.

There was a large gap between the subtree heights, so I knew there had to be more subtrees in the finalized state between those heights. It was very strange they weren't turning up.

So I checked the block height in the last sync status update, and noticed that the extra subtree was within 100 blocks of that height. So I guessed it was in the non-finalized state.

I also guessed it was in that state subtree list merge code because I wrote that code, I didn't remember limiting the number of subtrees after the list merge, and I don't remember testing for it either!

So let's add a test for that in this ticket, because I forgot to do that the first time around.
